src/ssort/_statements.py

```diff
@@ -12,53 +12,41 @@ class Statement:
     def __init__(
         self, *, text: str, node: ast.AST, start_row: int, start_col: int
     ) -> None:
-        self._text = text
-        self._node = node
-        self._start_row = start_row
-        self._start_col = start_col
+        self.text = text
+        self.node = node
+        self.start_row = start_row
+        self.start_col = start_col
+
+    @functools.cached_property
+    def text_padded(self) -> str:
+        """
+        Return the statement text padded with leading whitespace so that
+        coordinates in the ast match up with coordinates in the text.
+        """
+        return ("\n" * self.start_row) + (" " * self.start_col) + self.text
+
+    @functools.cached_property
+    def requirements(self) -> tuple[Requirement, ...]:
+        """
+        Returns an iterable yielding Requirement objects describing the
+        bindings that a statement references.
+        """
+        return tuple(get_requirements(self.node))
+
+    @functools.cached_property
+    def method_requirements(self) -> tuple[str, ...]:
+        """
+        Returns an iterable yielding the names of attributes of the `self`
+        parameter that a statement depends on.
+        """
+        return tuple(get_method_requirements(self.node))
+
+    @functools.cached_property
+    def bindings(self) -> tuple[str, ...]:
+        """
+        Returns an iterable yielding the names bound by a statement.
+        """
+        return tuple(get_bindings(self.node))
```

Comment:
I would prefer to keep these as iterators, because python 3.  Does the performance improvement go away if you create a `yield from` wrapper for a private cached property?

Reply:
I don't quite follow. A tuple is an iterator. We could of course return a generator over the cached tuple, but I don't see the benefit of doing so.

Reply:
Sorry for the confusion, in my use of the terminology a tuple is an iterable (it can be passed to `iter`), but not an iterator (it can't be passed to `next`).

Regarding tuple vs iterable vs iterator:  to me it's about reducing the interface to the minimum that makes sense.  My thought process is something like this:
  - Tuple supports arbitrary indexing which we don't need.
  - Iterable supports repeated iteration which we only need if the value is an attribute.
  - The initial access requires a non-trivial amount of computation, so the value should not be an attribute.
  - If the value is returned by a method then there is no need to support repeated iteration.
 
Although it's almost completely irrelevant here (a private, and carefully scrutinized API, with trivially auditable consumers) making the methods return an iterable rather than a tuple gives us more flexibility in how we implement the methods.

Reply:
Oops, you're right. I confused `Iterator` and `Iterable`. I'll make the change.